### DISCUSSION TITLE
FastAPI stops Accepting Requests after running for a while when deployed on Kubernetes

### DISCUSSION THREAD

**Discussion by NaumanHSA** [fastapi/fastapi#10229] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import uvicorn
from fastapi import FastAPI, BackgroundTasks, HTTPException, Request, status
from fastapi.middleware.cors import CORSMiddleware
from fastapi.responses import JSONResponse, HTMLResponse
from fastapi.encoders import jsonable_encoder
from fastapi.templating import Jinja2Templates
from fastapi.staticfiles import StaticFiles
from pydantic import BaseModel, Field, schema_json_of
import motor.motor_asyncio

logger, host_name = init.initLogger(logs_path=cfg.LOGS_PATH, log_level=cfg.LOG_LEVEL)
logger.info("Initializing DL Models...")
app = FastAPI()
app.mount("/logs", StaticFiles(directory="logs"), name="logs")
templates = Jinja2Templates(directory="templates")

origins = ["*"]
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

class OCRRequest(BaseModel):
    requestId: Optional[str] = Field(None)
    gt_data: Optional[dict] = Field(None)
    document: str 
    documentBack: str 
    save_local_response: Optional[bool] = Field(False)
    save_db_response: Optional[bool] = Field(True)
    vis: Optional[bool] = Field(False)


@app.post('/ocr')
async def ocr(request: OCRRequest, background_tasks: BackgroundTasks):
    try:
        requestId = request.requestId if request.requestId is not None else str(uuid.uuid4())
        logger.info(f"(ocr) [{requestId}] Execution Started-------------")
        SAVE_LOCAL_RESPONSE = request.save_local_response
        SAVE_DB_RESPONSE = request.save_db_response
        GT_DATA = request.gt_data
        SAVE_IMAGES = False
        VIS = request.vis

        ....
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=jsonable_encoder(err_res),
        )
```

### Description

Hi,

I've a containerized fastapi application which uses some machine learning models to do some processing. The response time for a single request is around 1.5 seconds with WEB_CONCURRENCY=2 (worker processes). I have deployed the solution on Kubernetes with 4 instances; 4 vCPUs and 8 GB memory each, and the estimated traffic is around 2000 requests per 24 hours per instance. When tested locally with the same configurations as the cloud, the application takes up to 1.5 GBs of memory.

fastapi==0.103.1
uvicorn==0.22.0

**Run**
I've docker file with the command below to run the application:
```
ENV WEB_CONCURRENCY=2
CMD ["uvicorn", "app:app", "--proxy-headers", "--host", "0.0.0.0", "--port", "80", "--timeout-keep-alive", "65"]
```

**Problem**
The problem I'm facing is after deploying the solution on AWS, it runs fine for a certain amount of time (3-4 hours), after which the instance status changes to **unhealthy** one by one, until all instances are unhealthy. After then, it will return 504 Gateway Timeout. Even though the pods are running in the backend, but ELB doesn't make it healthy. The health check URL is the base URL of the application returning status code 200, nothing fancy.
I've checked the logs the time instances are running. The logs show a linear flow of requests with linear response time (average 1.5 seconds) for each instance. See the logs below.

**Reason**
As per my understanding, the reason that the status changes to unhealthy is because the server is not responding due to some reasons, and the timeout of the ELB (60 seconds) finishes. 

**Experiments**
1. Tried the endpoint methods without async as default. The problem was that the memory usage would go on increasing linearly, resulting in the increase in response time. At one point, the instance would crash due to memory full.
2. Tried the endpoint methods with async which actually fixed the memory problem, now I get flat 1.5 GBs for one instance with 2 process workers. This also fixed the response time which is now around 1.5 seconds for a request.
3. Tries increasing the number of instances to 6.

I need help in understanding the behavior that the application stops responding which leads the instance to unhealthy. Is there any workaround for this?

Thanks in Advance !!!

![image](https://github.com/tiangolo/fastapi/assets/22982600/8bb0a430-1201-4837-b3ee-5a39ff5119f5)

### Operating System

Linux, Windows

### FastAPI Version

0.103.1

### Pydantic Version

1.10.8

### Python Version

3.8

**Comment by iudeen** [comment]
Do the following:

1. Setup liveliness probe and health probe for your pods. 
2. Run a single Uvicorn worker instead of 2 or more and scale the Kubernetes pod instead. 

We have been running with such setup in production for a while and haven't faced such issues.

**Comment by NaumanHSA** [comment]
Thank you @iudeen for the information. Switching to Gunicorn fixed the issue for me.
